### ISSUE TITLE
Instability with Long Sequences

### ISSUE THREAD

**Issue by FrancescaMiccolis** [lucidrains/titans-pytorch#29]
We are experiencing instability when training with long sequences, specifically with sequence lengths of 4096. The main issues we observe are:
The loss of the neural memory tends to diverge over time.
At some point during training, the predictions from the neural memory become NaN.
However, we found that setting max_grad_norm = 1.0 in the NeuralMemory helps prevent NaN values, at least for sequences of length 4096. Despite this, the issue of loss divergence remains.
Has anyone encountered similar problems, or are there known solutions/workarounds to improve stability when working with long sequences?
Any insights would be greatly appreciated.

**Comment by lucidrains** [comment]
@FrancescaMiccolis yea, i think that was sort of the elephant in the room for both the TTT paper as well this Titans. we are optimizing an inner network along with an outer one after all, so this is expected. besides experimenting with some of the normalization hyperparameters / mlp variants, at some point i want to try the feedforward proposed by the [normalized GPT](https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L292) out of nvidia and see if that makes a difference

**Comment by lucidrains** [comment]
@FrancescaMiccolis have you tried decreasing [this value](https://github.com/lucidrains/titans-pytorch/blob/main/train_mac.py#L42)?

**Comment by lucidrains** [comment]
another idea, again not in the paper, is to penalize the gradient norms a la gradient penalties from GAN literature.

**Comment by FrancescaMiccolis** [comment]
First of all, thank you for your availability and quick response!
we conducted further tests and observed that reducing the learning rate ( we used a learning rate of 0.001) prevents the model from going to NaN. However, we also noticed that it's generally better to disable memory rather than use it.
Let us know if you need any additional details

**Comment by lucidrains** [comment]
@FrancescaMiccolis that's surprising! i'll wait for your paper then 😄

**Comment by FrancescaMiccolis** [comment]
Thanks for your reply :) 

We are using the default parameters regarding the overlapping attention windows. 
We are not sure why memory doesn’t help in our case, but one possible reason could be the sequence length: during training, we use a segment length of 4096, while in testing, we keep the full sequence.

We’ll share more details as we investigate further!

**Comment by lucidrains** [comment]
@FrancescaMiccolis try turning off the overlapping windows and i think you'll see a clear difference with and without memory

it can be just that overlapping attention is way too strong for your task

**Comment by leonardoshen-ch** [comment]
Is Transformer still considered unbeatable, then?
